pkg/common/scheduler/README.md

```diff
@@ -0,0 +1,241 @@
+
+### Scheduler Profile
+
+#### Predicates:
+
+1. **PodFitsResources**: check if a pod has enough capacity [CORE]
+
+2. **NoMaxResourceCount**: check if total number of placement pods exceed available resources  [KAFKA]. It has an argument `NumPartitions` to configure the plugin with the total number of Kafka partitions.
+
+3. **EvenPodSpread**: check if resources are evenly spread across pods [CORE]. It has an argument `MaxSkew` to configure the plugin with an allowed skew factor.
+
+#### Priorities:
+
+1. **AvailabilityNodePriority**: make sure resources are evenly spread across nodes [CORE]. It has an argument `MaxSkew` to configure the plugin with an allowed skew factor.
+
+2. **AvailabilityZonePriority**: make sure resources are evenly spread across zones [CORE]. It has an argument `MaxSkew` to configure the plugin with an allowed skew factor.
+
+3. **LowestOrdinalPriority**: make sure vreplicas are placed on free smaller ordinal pods to minimize resource usage [CORE]
+
+**Example profile:**
+
+```
+data:
+  predicates: |+
+                  [
+                    {"Name": "PodFitsResources"},
+                    {"Name": "NoMaxResourceCount",
+                    "Args": "{\"NumPartitions\": 100}"},
+                    {"Name": "EvenPodSpread",
+                    "Args": "{\"MaxSkew\": 2}"}
+                  ]
+  priorities: |+
+                  [
+                    {"Name": "AvailabilityZonePriority",
+                    "Weight": 10,
+                    "Args":  "{\"MaxSkew\": 2}"},
+                    {"Name": "LowestOrdinalPriority",
+                    "Weight": 2}
+                  ]
+```
+
+### Descheduler Profile:
+
+#### Priorities:
+
+1. **RemoveWithAvailabilityNodePriority**: make sure resources are evenly spread across nodes [CORE]
+
+2. **RemoveWithAvailabilityZonePriority**: make sure resources are evenly spread across zones [CORE]
+
+3. **HighestOrdinalPriority**: make sure vreps are removed from higher ordinal pods to minimize resource usage [CORE]
+
+**Example profile:**
+
+```
+data:
+  priorities: |+
+                  [
+                    {"Name": "RemoveWithEvenPodSpreadPriority",
+                    "Weight": 10,
+                    "Args": "{\"MaxSkew\": 2}"},
+                    {"Name": "RemoveWithAvailabilityZonePriority",
+                    "Weight": 10,
+                    "Args":  "{\"MaxSkew\": 2}"},
+                    {"Name": "RemoveWithHighestOrdinalPriority",
+                    "Weight": 2}
+                  ]
+```
+
+### Recovery
+
+Some failure scenarios are described below:
+
+1. **Pod failure**:
+When a pod/replica in a StatefulSet goes down due to some reason (but its node and zone are healthy), a new replica is spun up by the StatefulSet with the same pod identit (pod can come up on a different node) almost immediately.
+All existing vreplica placements will still be valid and no rebalancing is needed.
+There shouldn’t be any degradation in Kafka message processing.
```

Comment:
This is not really true, a consumer group rebalance could degrade message processing especially when [Kafka Consumer Incremental Rebalance Protocol](https://cwiki.apache.org/confluence/display/KAFKA/KIP-429%3A+Kafka+Consumer+Incremental+Rebalance+Protocol) is not being used (which afaik is not implemented in Sarama).

Reply:
@pierDipi the pod set being referred to here is only talking about the eventing scheduler adapter pods where vreplicas are placed. Since pod will restart, the same placements can be kept without a rebalancing of the vreps. 
I agree with you about the consumer group rebalancing and degradation but that may/may not happen here if the kafka pods are affected, as well.
I hope I'm not missing anything...

Reply:
> @pierDipi the pod set being referred to here is only talking about the eventing scheduler adapter pods where vreplicas are placed. Since pod will restart, the same placements can be kept without a rebalancing of the vreps.

This is what `All existing vreplica placements will still be valid and no rebalancing is needed.` is saying, I agree and it's clear to me why.

I was referring to `There shouldn’t be any degradation in Kafka message processing.`.

> I agree with you about the consumer group rebalancing and degradation but that may/may not happen here if the kafka pods are affected, as well.
> I hope I'm not missing anything...

so, are you saying that if a pod where vreplicas are placed goes down that won't trigger a consumer group rebalance that affects message processing?

In the worst-case scenario, I'd expect something like this to happen (happy to be wrong):

1. Pod goes down
2. A new pod comes up (same name)
3. Kafka broker sees a new consumer that wants to join the group -> rebalance
4. Kafka detects that the consumer that was consuming messages in the dead pod (1) is not sending heartbeats anymore -> rebalance (again)

at least one rebalance happen. 2 in the worst case since `terminationGracePeriodSeconds = 0 < "time for Kafka to detect that a consumer is dead"`

Is the above not possible? If yes, does that count as a degradation for Kafka message processing?

Reply:
You're right. This is absolutely possible. 

I made an assumption that the same sticky pod (when restarted) would have the same consumer member ID and using static membership, would get the same assignment or so.

I don't have any numbers to quantify the extent of degradation for these recovery scenarios. Will need to run some performance runs to measure latency. Thank you for catching this @pierDipi !!